### DISCUSSION TITLE
Extended R dispatcher

### DISCUSSION THREAD

**Discussion by wlandau** [r-lib/mirai#78] [General]
@shikokuchuo, I finally had a free moment to try the extended R dispatcher you sketched in https://github.com/shikokuchuo/mirai/tree/dispatcher. Really promising stuff! Just writing in to share my initial thoughts and requests.

## Data

The `crew` launcher is an `R6` object. Is there a way to send this object to the `ext` argument along with the function and its call? I tried to ship data in the closure of the function, but in the example below, `f_data` does not go along for the ride.

```r
library(mirai)
f_data <- list2env(list(index = 0L), parent = emptyenv())
f <- function(envir = parent.frame()) {
  force(envir)
  tryCatch(
    f_data$index <- f_data$index + 1L,
    error = \(condition) writeLines(conditionMessage(condition), "error.txt")
  )
  saveRDS(as.list(envir), file.path("envir.rds"))
}
daemons(n = 2L, url = "ws://127.0.0.1:5700", ext = list(f, expression(f())))
task <- mirai(1)
daemons(0L)
```

## Launching

`crew` uses information about the current task backlog to decide if it should launch workers, as well as how many workers to launch. I think the extended R dispatcher is a great opportunity to simplify the logic of this. Is there a way to get the number of tasks that have not yet been dispatched? Also, the existing `serverfree` vector looks useful. Is it `FALSE` if the server is currently running a task and `TRUE` otherwise?

I ran some simple code to see my options:

```r
library(mirai)
f_data <- list2env(list(index = 0L), parent = emptyenv())
f <- function(envir = parent.frame()) {
  force(envir)
  saveRDS(as.list(envir), "envir.rds")
}
daemons(n = 2L, url = "ws://127.0.0.1:5700", ext = list(f, expression(f())))
urls <- nextget("urls")
lapply(urls, \(url) launch_local(url = url))
replicate(5L, mirai("task"))
daemons(0L)
str(readRDS("envir.rds")
```

Output below:

<details>

```r
List of 48
 $ free       : int(0) 
 $ changes    : logi [1:2] FALSE FALSE
 $ activevec  : int [1:2] 0 0
 $ cv_values  : int [1:2] 0 0
 $ cmessage   :Class 'recvAio' <environment: 0x12cbc6828> 
 $ sockc      :Classes 'nanoSocket', 'nano' <externalptr> 
  ..- attr(*, "id")= int 4
  ..- attr(*, "state")= chr "opened"
  ..- attr(*, "protocol")= chr "rep"
  ..- attr(*, "dialer")=List of 1
  .. ..$ :Classes 'nanoDialer', 'nano' <externalptr> 
  .. .. ..- attr(*, "id")= int 2
  .. .. ..- attr(*, "url")= chr "ipc:///tmp/9848fd9af4e9f66881f6f44ac"
  .. .. ..- attr(*, "state")= chr "started"
  .. .. ..- attr(*, "socket")= int 4
 $ ctrchannel : logi TRUE
 $ req        :Class 'recvAio' <environment: 0x12cbc5708> 
 $ ctx        :<externalptr> 
 $ listener   :Classes 'nanoListener', 'nano' <externalptr> 
  ..- attr(*, "id")= int 2
  ..- attr(*, "url")= chr "ws://127.0.0.1:5700/2"
  ..- attr(*, "state")= chr "started"
  ..- attr(*, "socket")= int 3
 $ ncv        :Class 'conditionVariable' <externalptr> 
 $ nsock      :Classes 'nanoSocket', 'nano' <externalptr> 
  ..- attr(*, "id")= int 3
  ..- attr(*, "state")= chr "opened"
  ..- attr(*, "protocol")= chr "req"
  ..- attr(*, "listener")=List of 1
  .. ..$ :Classes 'nanoListener', 'nano' <externalptr> 
  .. .. ..- attr(*, "id")= int 2
  .. .. ..- attr(*, "url")= chr "ws://127.0.0.1:5700/2"
  .. .. ..- attr(*, "state")= chr "started"
  .. .. ..- attr(*, "socket")= int 3
 $ nurl       : chr "ws://127.0.0.1:5700/2"
 $ burl       : chr "ws://127.0.0.1:5700/2"
 $ i          : int 2
 $ envir      :<environment: 0x12cbc1270> 
 $ ports      : NULL
 $ baseurl    : Named chr [1:10] "ws://127.0.0.1:5700" "ws" "" "127.0.0.1:5700" ...
  ..- attr(*, "names")= chr [1:10] "rawurl" "scheme" "userinfo" "host" ...
 $ active     :List of 2
  ..$ :Class 'conditionVariable' <externalptr> 
  ..$ :Class 'conditionVariable' <externalptr> 
 $ servers    :List of 2
  ..$ :Classes 'nanoSocket', 'nano' <externalptr> 
  .. ..- attr(*, "id")= int 2
  .. ..- attr(*, "state")= chr "opened"
  .. ..- attr(*, "protocol")= chr "req"
  .. ..- attr(*, "listener")=List of 1
  .. .. ..$ :Classes 'nanoListener', 'nano' <externalptr> 
  .. .. .. ..- attr(*, "id")= int 1
  .. .. .. ..- attr(*, "url")= chr "ws://127.0.0.1:5700/1"
  .. .. .. ..- attr(*, "state")= chr "started"
  .. .. .. ..- attr(*, "socket")= int 2
  ..$ :Classes 'nanoSocket', 'nano' <externalptr> 
  .. ..- attr(*, "id")= int 3
  .. ..- attr(*, "state")= chr "opened"
  .. ..- attr(*, "protocol")= chr "req"
  .. ..- attr(*, "listener")=List of 1
  .. .. ..$ :Classes 'nanoListener', 'nano' <externalptr> 
  .. .. .. ..- attr(*, "id")= int 2
  .. .. .. ..- attr(*, "url")= chr "ws://127.0.0.1:5700/2"
  .. .. .. ..- attr(*, "state")= chr "started"
  .. .. .. ..- attr(*, "socket")= int 3
 $ queue      :List of 2
  ..$ :List of 2
  .. ..$ ctx:<externalptr> 
  .. ..$ req:Class 'recvAio' <environment: 0x12cb72700> 
  ..$ :List of 2
  .. ..$ ctx:<externalptr> 
  .. ..$ req:Class 'recvAio' <environment: 0x12cbc5708> 
 $ serverfree : logi [1:2] TRUE TRUE
 $ activestore: int [1:2] 0 0
 $ instance   : int [1:2] 0 0
 $ complete   : int [1:2] 0 0
 $ assigned   : int [1:2] 0 0
 $ basenames  : chr [1:2] "ws://127.0.0.1:5700/1" "ws://127.0.0.1:5700/2"
 $ servernames: chr [1:2] "ws://127.0.0.1:5700/1" "ws://127.0.0.1:5700/2"
 $ seq_n      : int [1:2] 1 2
 $ vectorised : logi FALSE
 $ auto       : logi FALSE
 $ cv         :Class 'conditionVariable' <externalptr> 
 $ sock       :Classes 'nanoSocket', 'nano' <externalptr> 
  ..- attr(*, "id")= int 1
  ..- attr(*, "state")= chr "opened"
  ..- attr(*, "protocol")= chr "rep"
  ..- attr(*, "dialer")=List of 1
  .. ..$ :Classes 'nanoDialer', 'nano' <externalptr> 
  .. .. ..- attr(*, "id")= int 1
  .. .. ..- attr(*, "url")= chr "ipc:///tmp/9848fd9af4e9f66881f6f44a"
  .. .. ..- attr(*, "state")= chr "started"
  .. .. ..- attr(*, "socket")= int 1
 $ usecall    : logi FALSE
 $ usefunc    : logi TRUE
 $ call       :  expression(f())
 $ func       :function (envir = parent.frame())  
 $ host       : chr "ipc:///tmp/9848fd9af4e9f66881f6f44a"
 $ url        : chr "ws://127.0.0.1:5700"
 $ n          : int 2
 $ asyncdial  : logi FALSE
 $ token      : logi FALSE
 $ lock       : logi FALSE
 $ tls        : NULL
 $ pass       : NULL
 $ monitor    : chr "ipc:///tmp/9848fd9af4e9f66881f6f44ac"
 $ rs         : NULL
 $ ext        :List of 2
  ..$ :function (envir = parent.frame())  
  ..$ :  expression(f())
```

</details>

# Throttling

For performance, I would like to avoid calling the `crew` launcher every time the dispatcher wakes up. I was thinking of using `later` to implement a custom throttling mechanism: 

1. The first time the dispatcher wakes up, use `later` to schedule the `crew` launcher to run e.g. 0.25 seconds later.
2. If the dispatcher wakes up before those 0.25 seconds elapse, just go back to sleep.
3. Let the `later` main loop run the `crew` launcher at the 0.25-second mark.
4. Repeat (1) the next time the dispatcher wakes up.

I haven't experimented with this part just yet, but I was wondering if you see any pitfalls at this stage.

**Comment by wlandau** [comment]
Re throttling: I had hoped the example below would write a new file every second, but I see no files being written.

```r
library(mirai)
library(nanonext)

f <- function() {
  if (!is.finite(later::next_op_secs())) {
    later::later(
      func = \() writeLines(as.character(Sys.time()), basename(tempfile())),
      delay = 1
    )
  }
}

daemons(n = 2L, url = "ws://127.0.0.1:5700", ext = list(f, expression(f())))

now <- mclock()
while (mclock() - now < 3e4) {
  task <- mirai("task")
  msleep(10)
}

daemons(0L)
```

From the `later` README:

> To avoid bugs due to reentrancy, by default, scheduled operations only run when there is no other R code present on the execution stack; i.e., when R is sitting at the top-level prompt. 

That's probably the reason.

But maybe I can live without throttling if the launcher logic is fast enough. The original motivation was that in order to know how many new workers to launch, `crew` had to loop over all the tasks and check `nanonext::.unresolved()`. If instead `crew` can look up the number of undispatched tasks and use existing vectors like `online` and `serverfree`. Wouldn't suspect too much slowness from `serverfree & (online | (launching & !instance))` even if those vectors are thousands of elements long. (`launching` is a vector `crew` creates itself based on the startup time.)

**Reply by wlandau** [reply]
One thing I forgot about: rotation. How would you recommend rotating a server listener from inside the dispatcher? E.g. a replacement for `saisei()` to be invoked from the function in `ext`.

**Comment by shikokuchuo** [comment]
A couple of things:
- Currently 'ext' takes a list of a function and/or call. It's intended you pick either of depending on what's easier to use. You don't need f and expression(f) - you can specify one as NULL.
- Is it possible to define the R6 object inside your function (not in the closure)? I'll need to take a closer look if there's a reason this doesn't work.
- Running later on dispatcher should not be necessary. Also as we're waiting it'll never run...
- `(activevec & serverfree)` I think that gets you the available workers i.e. connected and free. There should be enough logic not to call unresolved().
- To rotate listeners, perhaps copy these lines of code for now https://github.com/shikokuchuo/mirai/blob/7640107249dc2018c8502310cde3c8d787aa6026/R/mirai.R#L326-L330. This is a good point - eventually I will need to make an interface like `nextget()` for all the functions you call, also possibly the other variables so that this can remain stable.

**Reply by wlandau** [reply]
> Is it possible to define the R6 object inside your function (not in the closure)? I'll need to take a closer look if there's a reason this doesn't work.

Unfortunately not, because the `R6` object is a mutable static variable, and its state needs to carry over from each call to the next.  Also, it is best if this object can be created on the host so the `validate()` method can run and report user errors more easily. 

> Running later on dispatcher should not be necessary. Also as we're waiting it'll never run...

It shouldn't be strictly necessary, I was just thinking of efficiency. In `crew`, I have observed slowness with this pattern:

1. Submit one task.
2. Take the time to figure out how many workers to launch.
3. Launch only one worker.
4. Submit another task.
5. Take the time to figure out how many workers to launch.
6. Launch only one worker.

With `crew`'s current polling-based throttling mechanism, responsiveness improves:

1. Submit a group of tasks.
2. Take the time to figure out how many workers to launch.
3. Launch a bunch of workers at once.

> (activevec & serverfree) I think that gets you the available workers i.e. connected and free. There should be enough logic not to call unresolved().

What exactly do `activevec` and `serverfree` mean? Do they just pertain to `mirai` servers? If so, then `crew` also needs information about the task load in order to know how many new worker launches are necessary. Server-only information is unfortunately not enough.

> To rotate listeners, perhaps copy these lines of code for now 

Thanks, that will help me get started. The only other things I think I will need are the ability to send a static `R6` object and the ability to get task-specific information (preferably the number of tasks not yet dispatched or queued to be re-sent).

**Reply by shikokuchuo** [reply]
OK great!

1. I think I understand the delayed launch problem - let me think about it - it feels like there should be a solution.
2. For the R6 object - a `daemons(dispatcher = TRUE, ext = crew::dispatcher_extension)` is how I imagined it would work. The only thing being passed to dispatcher is a text argument. You want to instantiate an R6 object before passing it to dispatcher - that sounds like it would have to be sent, correct?
3. `activevec` are the servers that are connected, and `serverfree` are the servers which can accept new tasks, so both are server statistics. Tasks queued, presumably you had that on host, but need that on dispatcher.

At the moment, it seems 1 and 3 may be solved by passing information to dispatcher through the existing control channel. Would `crew()` call `status()` often enough for this to work (so we would send information from host to dispatcher as well as receiving from dispatcher to host), or is the idea to be even more efficient?

**Reply by wlandau** [reply]
> For the R6 object - a daemons(dispatcher = TRUE, ext = crew::dispatcher_extension) is how I imagined it would work. The only thing being passed to dispatcher is a text argument. You want to instantiate an R6 object before passing it to dispatcher - that sounds like it would have to be sent, correct?

Yes, exactly. And if its internal state could be preserved and updated from call to call, that will allow `crew` to correctly track the workers it has already launched. (As well as other information such as when launches occurred, so as to allow an appropriate startup time window.)  

> activevec are the servers that are connected,

Same as the `online` vector?

> `serverfree` are the servers which can accept new tasks

Does that mean `serverfree` is `TRUE` if a server is connected and not running a task? Would it be `FALSE` if a server is disconnected or there is a task in progress?

> Tasks queued, presumably you had that on host, but need that on dispatcher.

Yes, currently I get this information by inefficiently looping over `nanonext::.unresolved()` over all the thousands of `mirai` objects every 0.25 seconds. If instead there were a simple counter on the dispatcher which just counts the number of tasks that have not been dispatched yet, I could cut out this slow `.unresolved()` loop entirely, and the logic on my end would be a lot simpler too. But I agree, there are multiple ways to get the kind of task information that `crew` would need.

> Would crew() call status() often enough for this to work (so we would send information from host to dispatcher as well as receiving from dispatcher to host), or is the idea to be even more efficient?

If I understand correctly, then it seems like this may defeat the purpose. The host would have to call `.unresolved()` on every task and then send the count  back to the dispatcher. To make sure the `.unresolved()` count is up to date, this would require frequent  polling on the host, which is what `ext` is trying to avoid in the first place.

(...unless that count could automatically increment on a call to `mirai::mirai()` and automatically decrement on the host without having to call `.unresolved()` synchronously, but the latter seems like magic...)

And the `.unresolved()` count isn't all that nice anyway because the logic is complicated and the inferential steps may not always be strictly true. What `crew` is really trying to do is figure out the number of tasks in the queue (prior to dispatch), then launch any workers necessary to run those tasks (up to a sensible limit). The current logic tries to get at this backlog task count indirectly, and it does an okay job, but `crew` would be more seamless if this count could be determined more directly.

**Reply by shikokuchuo** [reply]
Thanks this is all very useful. I think I may have already implemented the solution to your `.unresolved()` problem. Just give me a little more time and I will let you know.

**Reply by shikokuchuo** [reply]
> > activevec are the servers that are connected,
> 
> Same as the `online` vector?

Yes, except 'online' doesn't exist on dispatcher, there it is still 'activevec'.

> 
> > `serverfree` are the servers which can accept new tasks
> 
> Does that mean `serverfree` is `TRUE` if a server is connected and not running a task? Would it be `FALSE` if a server is disconnected or there is a task in progress?

'serverfree' is TRUE only if it can accept a task immediately. There are multiple reasons it might be FALSE - either busy, or the server is exiting. As the extension is running at the end of the loop, just before it waits again, this means that if there is a backlog (tasks >= servers) then this vector would all be FALSE. I hope this bit of logic helps even if you don't know the full backlog (yet).

**Reply by wlandau** [reply]
> 'serverfree' is TRUE only if it can accept a task immediately. There are multiple reasons it might be FALSE - either busy, or the server is exiting. As the extension is running at the end of the loop, just before it waits again, this means that if there is a backlog (tasks >= servers) then this vector would all be FALSE. I hope this bit of logic helps even if you don't know the full backlog (yet).

That's really helpful. So if any of `serverfree` is `TRUE`, then all tasks have already been successfully dispatched, and `crew` does not need to launch more servers/workers. Otherwise, all of `serverfree` is `FALSE` all active servers are busy doing something (running or exiting).

If all of `serverfree` is `FALSE`, then might be a backlog of `n` tasks that could not be dispatched. In this scenario, all the "online"/"activevec" workers are busy and cannot accept any of these `n` tasks. However, there may be `m` workers which launched recently but did not connect yet. (`m` is equal to `sum(launching & !instance)`.) We should give these `m` workers more time (a total of `seconds_launch` in `crew` parlance) to boot up, connect to their NNG sockets, and accept backlogged tasks. For the current call to the dispatcher, all this boils down to a decision: `crew` should launch `n - m` new worker instances (actually, `max(length(nextget("urls")), min(n - m, 0L))` new worker instances: these bounds avoid submitting too many workers or thinking in terms of negative workers). The launches should be directed towards freshly-rotated sockets for which instances are not already launching or running. I hope that helps.

**Reply by shikokuchuo** [reply]
> Thanks this is all very useful. I think I may have already implemented the solution to your `.unresolved()` problem. Just give me a little more time and I will let you know.

I refer to the additions as part of this commit 0364a1a, which adds signalling 'mirai'. I was thinking that if you just call `nanonext::cv_value()` on `nextget("cv")`, this will increase each time a mirai resolves so you don't need to manually query all of them. There is also the option of waiting on a resolution (as opposed to polling).

This may require a refactor of how things work at the moment though as I see you 'collect' the resolved ones, but I am not sure if that is strictly necessary. If you can just find a resolved one when you pop then you could make use of this. Note that if you do use signalling mirai, you will have to use regular `resolved()` rather than `.unresolved()` as `.unresolved()` does not know how to acquire the mutex lock within the cv.

But in any case, have a look at the changelog for all the recent changes - it's turning out to be quite a lot! I think I will look to make a CRAN release to consolidate the changes thus far as there is clearly much more work for the extension to become viable.

**Reply by shikokuchuo** [reply]
Btw my previous comment was more by way of observation rather than saying anything needs to change.

I need to take the time to generally update and improve documentation for mirai as well as actually writing those vignettes, so ideally we can catch up on this some time next week once that's done.

**Reply by wlandau** [reply]
`cv_value(nextget("cv"))` is an interesting idea, and I think it could potentially simplify and speed up the current polling-based version of `crew`. `cv_value(nextget("cv"))` counts the resolved tasks, and `controller$push()` could maintain an internal count of the launched ones. That's enough information for auto-scaling with polling-based `crew`, and it would remove the need for `.unresolved()` in this scenario.

As for waiting on `nextget("cv")`: that's a really easy and tempting way to approach https://github.com/wlandau/crew/issues/108. But for polling-based `crew`, `crew` would also need to be sensitive to worker crashes, etc.

With the hypothetical non-polling-based `crew` proposed in this thread, `nextget("cv")` seems like it would get us part of the way there. It's an awesome start! However, for auto-scaling purposes, `crew` needs to count the *unresolved* tasks, not the resolved ones. In addition, it would need to access these condition variables from inside the function in `ext` (more precisely, the pre-initialized `R6` launcher object sent to the dispatcher through `ext`). So if `mirai(.signal = TRUE)` could increment a `cv_submitted` variable immediately on task submission (not sure if this exists yet) and then a `cv_resolved` variable when it is resolved (which is what you just cleverly implemented), then the `crew` launcher in `ext` will be able to count the unresolved tasks. With these condition variables available in the dispatcher, along with persistent (`R6`) objects and a stable `ext`-compatible `saisei()`, `crew` will be in an excellent position to make use of the `mirai` dispatcher for launching workers.

**Reply by shikokuchuo** [reply]
Let me park the non-polling issues to next week as I'll need to spend some proper time on them.

But to answer your earlier ones:

> cv_value(nextget("cv")) is an interesting idea, and I think it could potentially simplify and speed up the current polling-based version of crew. cv_value(nextget("cv")) counts the resolved tasks, and controller$push() could maintain an internal count of the launched ones. That's enough information for auto-scaling with polling-based crew, and it would remove the need for .unresolved() in this scenario.

If this works, then you are more than welcome to implement it - I thought it might but wasn't sure. The signalling mirai is here to stay as it was introduced to achieve the synchronous load-balancing behaviour for `parallel` clusters.

> As for waiting on nextget("cv"): that's a really easy and tempting way to approach https://github.com/wlandau/crew/issues/108. But for polling-based crew, crew would also need to be sensitive to worker crashes, etc.

~~Guarding against crashes was also a consideration for the parallel backend, so this I solved by setting a pipe_notify event on the socket - `nextget("sock")` - this interface is coming in very handy now! Essentially this line: https://github.com/shikokuchuo/mirai/blob/b7da122a2ac7ff00c2ddb1198571610a0db9f5b5/R/parallel.R#L115~~

~~This will signal the cv on disconnect so that any waits are safe. Also by setting the flag, if it wakes from a disconnect as opposed to any other event, it will return FALSE instead of TRUE, hence allowing handling of the type below:~~
https://github.com/shikokuchuo/mirai/blob/b7da122a2ac7ff00c2ddb1198571610a0db9f5b5/R/parallel.R#L177-L180

Update: sorry the idea doesn’t work as crew has semi-transient workers which connect and disconnect all the time.

**Reply by wlandau** [reply]
> Update: sorry the idea doesn’t work as crew has semi-transient workers which connect and disconnect all the time.

Yeah, this is turning out to be really tricky. 

This weekend, I even thought of an edge case that might make the extended R dispatcher infeasible for `crew`. Suppose a user pushes a task, the dispatcher wakes up, and then the `crew` launcher in `ext` tries to start a worker on an AWS Batch spot instance. If the spot instance fails to start because the user's price threshold is not reached within the startup window, then the worker will not connect to its NNG websocket. If that happens, then the dispatcher will sleep through the failure, and nothing will be able to trigger a retry.

**Reply by wlandau** [reply]
I hope this isn't too discouraging, I know you already put work into https://github.com/shikokuchuo/mirai/tree/dispatcher. I wish I had thought of https://github.com/shikokuchuo/mirai/discussions/78#discussioncomment-7164289 sooner.

**Reply by wlandau** [reply]
I guess, if the dispatcher were to have light-touch polling, even as infrequently as once per second, I might be able to work with that. Not sure how messy or strange that would be on your end, and I totally understand if you don't want to go there.

**Reply by shikokuchuo** [reply]
> > Update: sorry the idea doesn’t work as crew has semi-transient workers which connect and disconnect all the time.
> 
> Yeah, this is turning out to be really tricky.
> 
> This weekend, I even thought of an edge case that might make the extended R dispatcher infeasible for `crew`. Suppose a user pushes a task, the dispatcher wakes up, and then the `crew` launcher in `ext` tries to start a worker on an AWS Batch spot instance. If the spot instance fails to start because the user's price threshold is not reached within the startup window, then the worker will not connect to its NNG websocket. If that happens, then the dispatcher will sleep through the failure, and nothing will be able to trigger a retry.

Interesting case. I think the solution here is to use a `later` poll in crew on the host process, not from dispatcher. So it will keep checking until the connection happens - if it doesn't then it can trigger dispatcher to run again etc. The key is that you would need to trust `later` to run in the background - R has to be at the top level prompt for this to happen.

**Reply by shikokuchuo** [reply]
> I hope this isn't too discouraging, I know you already put work into https://github.com/shikokuchuo/mirai/tree/dispatcher. I wish I had thought of [#78 (reply in thread)](https://github.com/shikokuchuo/mirai/discussions/78#discussioncomment-7164289) sooner.

Not at all discouraging, to be honest I haven't thought through the point about the R6 object needing to be passed through to dispatcher yet. If it proves too difficult, we can always try to jump straight to threaded dispatcher.

**Reply by wlandau** [reply]
> Interesting case. I think the solution here is to use a later poll in crew on the host process, not from dispatcher. So it will keep checking until the connection happens - if it doesn't then it can trigger dispatcher to run again etc. The key is that you would need to trust later to run in the background - R has to be at the top level prompt for this to happen.

If only `later` could work inside functions and had the power to suspend the main event loop to run a quick task. Then the host process would be perfect and `ext` would have been moot from the beginning. Unfortunately, `later` conflicts with the major non-interactive use cases of `crew`, including `targets`. 

> Not at all discouraging, to be honest I haven't thought through the point about the R6 object needing to be passed through to dispatcher yet. If it proves too difficult, we can always try to jump straight to threaded dispatcher.

That's good to know. It's starting to look really challenging:

1. Special task-level condition variables on the dispatcher.
2. `R6` objects in `ext`.
3. `saisei()` on the dispatcher.
4. How to detect crashes that happen prior to connecting.

All this makes me think the threaded dispatcher would be much more feasible. It's a challenge in and of itself, but at least `crew` is far less likely to add additional burdens like the 4 above.

**Reply by wlandau** [reply]
By the way, I started using `nextget("cv")` in `crew`, and it simplified so much that the major challenge was deprecating arguments and methods! `crew` is simpler than it has ever been, and design elements are falling into place so well. Which now makes me hesitate even more when it comes to using `ext` in `crew`.

**Reply by shikokuchuo** [reply]
Wow your `crew` backlog tests seem to run a *lot* faster!

**Reply by wlandau** [reply]
Yeah, I am really enjoying that!

Also, an update: I don't think I can entirely avoid `.unresolved()` just because of how `controller$wait(mode = "one")` works, but I no longer need it for auto-scaling.
